src/coreclr/src/jit/codegenxarch.cpp

```diff
@@ -3005,20 +3011,36 @@ void CodeGen::genCodeForInitBlkUnroll(GenTreeBlk* node)
             // For x86, we need one more to convert it from 8 bytes to 16 bytes.
             emit->emitIns_R_R(INS_punpckldq, EA_16BYTE, srcXmmReg, srcXmmReg);
 #endif
+            if (useYmm)
+            {
+                emit->emitIns_R_R(INS_punpckldq, EA_32BYTE, srcXmmReg, srcXmmReg);
+            }
         }
 
         instruction simdMov = simdUnalignedMovIns();
-        for (unsigned regSize = XMM_REGSIZE_BYTES; size >= regSize; size -= regSize, dstOffset += regSize)
-        {
-            if (dstLclNum != BAD_VAR_NUM)
-            {
-                emit->emitIns_S_R(simdMov, EA_ATTR(regSize), srcXmmReg, dstLclNum, dstOffset);
-            }
-            else
+
+        auto unrollUsingXMM = [&](unsigned regSize) {
+            for (; size >= regSize; size -= regSize, dstOffset += regSize)
             {
-                emit->emitIns_ARX_R(simdMov, EA_ATTR(regSize), srcXmmReg, dstAddrBaseReg, dstAddrIndexReg,
-                                    dstAddrIndexScale, dstOffset);
+                if (dstLclNum != BAD_VAR_NUM)
+                {
+                    emit->emitIns_S_R(simdMov, EA_ATTR(regSize), srcXmmReg, dstLclNum, dstOffset);
+                }
+                else
+                {
+                    emit->emitIns_ARX_R(simdMov, EA_ATTR(regSize), srcXmmReg, dstAddrBaseReg, dstAddrIndexReg,
+                                        dstAddrIndexScale, dstOffset);
+                }
             }
+        };
+
+        if (useYmm)
+        {
+            unrollUsingXMM(YMM_REGSIZE_BYTES);
+        }
+        if (size >= XMM_REGSIZE_BYTES)
+        {
+            unrollUsingXMM(XMM_REGSIZE_BYTES);
```

Comment:
This is to handle trailing elements, which should be at most one store, right?

Reply:
Yes, if YMM are used, otherwise it will be `size/XMM_REGSIZE_BYTES` times.